lib/puppet/http/service/compiler.rb

```diff
@@ -129,6 +129,68 @@ def post_catalog(name, facts:, environment:, configured_environment: nil, transa
+  # @param [Hash] persistent A hash containing two required keys, facts and catalog,
+  #   which when set to true will cause the facts and reports to be stored in
+  #   PuppetDB, or discarded if set to false.
+  # @param [String] environment The name of the environment for which to compile the catalog.
+  # @param [Hash] facts A hash with a required values key, containing a hash of all the
+  #    facts for the node. If not provided, Puppet will attempt to fetch facts for the node
+  #    from PuppetDB.
+  # @param [Hash] trusted_facts A hash with a required values key containing a hash of
+  #    the trusted facts for a node
+  # @param [String] transaction_uuid The id for tracking the catalog compilation and
+  #    report submission.
+  # @param [String] job_id The id of the orchestrator job that triggered this run.
+  # @param [Hash] options A hash of options beyond direct input to catalogs. Options:
+  #    - prefer_requested_environment Whether to always override a node's classified 
+  #      environment with the one supplied in the request. If this is true and no environment
+  #      is supplied, fall back to the classified environment, or finally, 'production'.
+  #    - capture_logs Whether to return the errors and warnings that occurred during
+  #      compilation alongside the catalog in the response body.
+  #    - log_level The logging level to use during the compile when capture_logs is true.
+  #      Options are 'err', 'warning', 'info', and 'debug'.
+  #
+  # @return [Array<Puppet::HTTP::Response, Puppet::Resource::Catalog, Array<String>>] An array
+  #   containing the request response, the deserialized catalog returned by
+  #   the server and array containing logs (log array will be empty if capture_logs is false)
+  #
+  def post_catalog4(certname, persistence, environment, facts: nil, trusted_facts: nil, transaction_uuid: nil, job_id: nil, options: nil)
+    body = {
+      certname: certname,
+      persistence: persistence,
+      environment: environment,
+      transaction_uuid: transaction_uuid,
+      job_id: job_id,
+      options: options
+    }
+    body[:facts] = { values: facts } unless facts.nil?
+    body[:trusted_facts] = { values: trusted_facts } unless trusted_facts.nil?
+    headers = add_puppet_headers(
+      'Accept' => get_mime_types(Puppet::Resource::Catalog).join(', '),
+      'Content-Type' => 'application/json'
+    )
+
+    url = URI::HTTPS.build(host: @url.host, port: @url.port, path: Puppet::Util.uri_encode("/puppet/v4/catalog"))
+    response = @client.post(
+      url,
+      body.to_json,
+      headers: headers
+    )
+    process_response(response)
+    raise Puppet::HTTP::ResponseError.new(response) unless response.success?
+    ## CODEREVIEW Should we do anything to catch a json deserialization error?
+    response_body = JSON.parse(response.body)
+    catalog = Puppet::Resource::Catalog.from_data_hash(response_body['catalog'])
```

Comment:
Yeah if `JSON.parse` or `from_data_hash` fails we should raise a `Puppet::HTTP::SerializationError` like we do in https://github.com/puppetlabs/puppet/blob/b9240c317780df0359eb2451db0a4cb0177a353a/lib/puppet/http/service.rb#L164

Reply:
thanks, updated and added unit test.

Reply:
oh, wait i just realized i didnt do anything about handling failures in from_data_hash. Will update.

Reply:
Updated